### PR TITLE
BUG: Ensure correct handling of multidimensional q by nanquantile

### LINKED ISSUE #25731: BUG: Nanquantile doens't handle multidimensional `q` correctly
Quantile/percentile insert the dimensions of `q` at the beginning of the result.  Nanquantile seems not to be written to correctly support multi-dimensional `q` and ends up splitting it up:
```python
a = np.ones((3, 4))
q = np.full((1, 2), 0.5)
 
print(np.quantile(a, q, axis=1).shape)
# (1, 2, 3)  correct
print(np.nanquantile(a, q, axis=1).shape)
# (1, 3, 2) mangled up
```
The axes contributed by `q` of shape `(1, 2)` should end up at the beginning but get split.

(Reported by @aschaffer)

### PR DESCRIPTION
**Pull request by amoghsood**
Fixes #25731.
Key changes applied to the following function `def _nanquantile_ureduce_func()` in `/numpy/lib/_nanfunctions_impl.py`
In particular the results from `result = _nanquantile_1d(part, q, overwrite_input, method, weights=wgt)` were not being reshaped properly. Tests passed on local plus contributed additional test to `numpy/lib/tests/test_nanfunctions.py` to validate consistent behavior between quantile and nanquantile. 
Some additional details included below to facilitate code-review.
Original behavior that was highlighted in #25731.

```
a = np.ones((3, 4))
q = np.full((1, 2), 0.5)
 
print(np.quantile(a, q, axis=1).shape)
# (1, 2, 3)  correct
print(np.nanquantile(a, q, axis=1).shape)
# (1, 3, 2) mangled up
```
To ensure consistent behavior with `np.quantile/np.percentile` the axes contributed by `q` should end up at the beginning of the result. 

The key issue was `def _nanquantile_ureduce_func()` in `/numpy/lib/_nanfunctions_impl.py`.  All the computations seem correct (in my understanding), up untill `result = _nanquantile_1d(part, q, overwrite_input, method, weights=wgt)`. After this, the `result` was being reshaped using `result = np.moveaxis(result, axis, 0)`. This resulted in the mangled up shape highlighted in the original bug.

The proposed solution explicitly computes the locations of axis, before applying `np.moveaxis`
```
num_dims = len(a.shape)+len(q.shape)-1
new_axes = list(range(len(q.shape), num_dims))
new_axes[axis:axis] = range(len(q.shape)) 
result = np.moveaxis(result, source=range(num_dims), destination=new_axes)
```
Now, the return values and shapes of `np.nanquantile` are consistent with `np.quantile`. This has been verified using newly contributed tests included in `test_consistency_with_quantile` in `numpy/lib/tests/test_nanfunctions.py`. Furthermore, all pre-existing tests pass.

### CODE REVIEW

File: numpy/lib/tests/test_nanfunctions.py [hunk 1]
```diff
@@ -1265,6 +1265,14 @@ def test_regression(self, weighted):
...
         assert_equal(np.nanquantile(ar, q=[0.25, 0.5, 0.75], axis=1, **w_args),
                      np.nanpercentile(ar, q=[25, 50, 75], axis=1, **w_args))
+    
+    @pytest.mark.parametrize(argnames='axis', 
+        argvalues=[1, (1, ), 2, (0, 2), -1, (-1, -2)])
+    def test_consistency_with_quantile(self, axis):
+        a = np.arange(42).reshape(3, 7, 2).astype(float)
+        q = np.full((3, 2), 0.5)
+        assert_equal(np.nanquantile(a, q, axis=axis).shape, 
+                      np.quantile(a, q, axis=axis).shape)
```
Comment (lorentzenchr) [inline numpy/lib/tests/test_nanfunctions.py:1275]: Could add an assert that test for equal values, not only the shapes?

File: numpy/lib/tests/test_nanfunctions.py [hunk 2]
```diff
@@ -1265,6 +1265,14 @@ def test_regression(self, weighted):
                      np.nanpercentile(ar, q=[50], axis=1, **w_args))
         assert_equal(np.nanquantile(ar, q=[0.25, 0.5, 0.75], axis=1, **w_args),
                      np.nanpercentile(ar, q=[25, 50, 75], axis=1, **w_args))
+    
+    @pytest.mark.parametrize(argnames='axis', 
+        argvalues=[1, (1, ), 2, (0, 2), -1, (-1, -2)])
+    def test_consistency_with_quantile(self, axis):
+        a = np.arange(42).reshape(3, 7, 2).astype(float)
```
Comment (lorentzenchr) [inline numpy/lib/tests/test_nanfunctions.py:1272]: There are different code path for `a.ndim==1` and `a.ndim > 1`, so could parametrize the shape to cover this?

File: numpy/lib/_nanfunctions_impl.py [hunk 3]
```diff
@@ -1646,14 +1646,20 @@ def _nanquantile_ureduce_func(
...
         result = np.apply_along_axis(_nanquantile_1d, axis, a, q,
                                      overwrite_input, method, weights)
         # apply_along_axis fills in collapsed axis with results.
         # Move that axis to the beginning to match percentile's
         # convention.
         if q.ndim != 0:
-            result = np.moveaxis(result, axis, 0)
-
+            #GH-25731
+            num_dims = len(a.shape)+len(q.shape)-1
```
Comment (lorentzenchr) [inline numpy/lib/_nanfunctions_impl.py:1658]:
```suggestion
            num_dims = a.ndim + q.ndim - 1
```
Same for the other occurrences of `len(array.shape)`.

File: numpy/lib/_nanfunctions_impl.py [hunk 4]
```diff
@@ -1646,14 +1646,20 @@ def _nanquantile_ureduce_func(
...
+            new_axes = list(range(len(q.shape), num_dims))
```
Comment (lorentzenchr) [inline numpy/lib/_nanfunctions_impl.py:1659]:
```suggestion
            new_axes = list(range(0, q.ndim))
```
Shouldn't this do the trick? We want the dimensions of q to be the first ones.

File: numpy/lib/tests/test_nanfunctions.py (hunk 2 above)
Comment (amoghsood) [reply numpy/lib/tests/test_nanfunctions.py:1272]: @lorentzenchr Thank you for your helpful comments. I reckon the following should give adequate coverage? What do you think ? Based on your suggestions I propose adding assert that checks for equality alongside shapes and have parametrized over shape as well.
```
@pytest.mark.parametrize(argnames='axis',
     argvalues=[None, 0, 1, (1, ), 2, (0, 2), -1, (-1, -2)])
@pytest.mark.parametrize(argnames='shape',
                         argvalues=[(42, ), (3, 7, 2)])
def test_consistency_with_quantile(self, shape, axis):
    a = np.arange(np.prod(shape)).reshape(shape).astype(float)
    if a.ndim == 1 and axis not in (None, 0, -1):
        pytest.skip(f"Skipping test for {a.ndim}-D array with incompatible axis {axis}")
    q = np.full((3, 2), 0.5)
    assert_equal(np.nanquantile(a, q, axis=axis).shape,
                  np.quantile(a, q, axis=axis).shape)
    assert_equal(np.nanquantile(a, q, axis=axis),
                  np.quantile(a, q, axis=axis))
```

File: numpy/lib/tests/test_nanfunctions.py [hunk 5]
```diff
@@ -1266,6 +1266,21 @@ def test_regression(self, weighted):
...
                      np.nanpercentile(ar, q=[25, 50, 75], axis=1, **w_args))
 
+    @pytest.mark.parametrize(argnames='axis',
+         argvalues=[None, 0, 1, (1, ), 2, (0, 2), -1, (-1, -2)])
+    @pytest.mark.parametrize(argnames='shape',
+                             argvalues=[(42, ), (3, 7, 2)])
+    def test_consistency_with_quantile(self, shape, axis):
+        a = np.arange(np.prod(shape)).reshape(shape).astype(float)
+        if a.ndim == 1 and axis not in (None, 0, -1):
+            pytest.skip(f"Skipping test for {a.ndim}-D array with incompatible axis {axis}")
```
Comment (charris) [inline numpy/lib/tests/test_nanfunctions.py:1276]: Please break this line.

File: numpy/lib/_nanfunctions_impl.py (hunk 3 above)
Comment (amoghsood) [reply numpy/lib/_nanfunctions_impl.py]: @lorentzenchr Thank you for your suggestions. Definitely helped improve clarity and readability.

File: numpy/lib/_nanfunctions_impl.py (hunk 4 above)
Comment (amoghsood) [reply numpy/lib/_nanfunctions_impl.py]: @lorentzenchr I am pretty sure I am misunderstanding your proposed solution. However just changing `new_axes` in the place you proposed will result in a value error `ValueError: repeated axis in `destination` argument` in `np.moveaxis()`.  (As a concrete example I consider, `a = np.ones((3, 4))`
`q = np.full((1, 2), 0.5)` and `np.nanquantile(a, q, axis=1).shape)`)

File: numpy/lib/_nanfunctions_impl.py [hunk 6]
```diff
@@ -1646,14 +1646,20 @@ def _nanquantile_ureduce_func(
         wgt = None if weights is None else weights.ravel()
         result = _nanquantile_1d(part, q, overwrite_input, method, weights=wgt)
     else:
+        #GH-25731
```
Comment (lorentzenchr) [inline numpy/lib/_nanfunctions_impl.py:1649]: I would replace the gh-xxx by a short comment to ease reading the code - or remove the comment entirely.

File: numpy/lib/_nanfunctions_impl.py [hunk 7]
```diff
@@ -1646,14 +1646,20 @@ def _nanquantile_ureduce_func(
...
+        axis = _nx.normalize_axis_tuple(axis, a.ndim)[0]
         result = np.apply_along_axis(_nanquantile_1d, axis, a, q,
                                      overwrite_input, method, weights)
         # apply_along_axis fills in collapsed axis with results.
         # Move that axis to the beginning to match percentile's
         # convention.
         if q.ndim != 0:
-            result = np.moveaxis(result, axis, 0)
-
+            #GH-25731
```
Comment (lorentzenchr) [inline numpy/lib/_nanfunctions_impl.py:1657]: Same here, a real comment or rather no comment.

File: numpy/lib/_nanfunctions_impl.py [hunk 8]
```diff
@@ -1646,14 +1646,20 @@ def _nanquantile_ureduce_func(
...
+            num_dims = a.ndim + q.ndim - 1
+            new_axes = list(range(q.ndim, num_dims))
+            new_axes[axis:axis] = range(q.ndim)
```
Comment (lorentzenchr) [inline numpy/lib/_nanfunctions_impl.py:1660]: Is the slicing `axis:axis` necessary?

File: numpy/lib/tests/test_nanfunctions.py [hunk 9]
```diff
@@ -1266,6 +1266,22 @@ def test_regression(self, weighted):
...
+    def test_consistency_with_quantile(self, shape, axis):
+        a = np.arange(np.prod(shape)).reshape(shape).astype(float)
+        if a.ndim == 1 and axis not in (None, 0, -1):
+            pytest.skip(f"Skipping test for {a.ndim}-D array"
+                        f" with incompatible axis {axis}")
+
+        q = np.full((3, 2), 0.5)
+        assert_equal(np.nanquantile(a, q, axis=axis).shape,
+                     np.quantile(a, q, axis=axis).shape)
+        assert_equal(np.nanquantile(a, q, axis=axis),
```
Comment (lorentzenchr) [inline numpy/lib/tests/test_nanfunctions.py:1282]: One single `assert_array_equal(…, strict=True` would be enough.

**Review by lorentzenchr — COMMENTED** [review]
A few last comments from my side.
For a numpy dev, there might be an easier or more obvious way to handle the dimensions that I just don’t know about.

**Comment by seberg** [comment]
Sorry that this was stalled, it looked pretty close.  But I recently had to fix another thing and fixed this incidentally.

**Pull request closed without merging**